kernel/src/table_features/column_mapping.rs

```diff
@@ -25,18 +29,135 @@ pub(crate) fn column_mapping_mode(
     protocol: &Protocol,
     table_properties: &TableProperties,
 ) -> ColumnMappingMode {
-    match table_properties.column_mapping_mode {
-        Some(mode) if protocol.min_reader_version() == 2 => mode,
-        Some(mode)
-            if protocol.min_reader_version() == 3
-                && protocol.has_reader_feature(&ReaderFeatures::ColumnMapping) =>
-        {
-            mode
-        }
+    match (
+        table_properties.column_mapping_mode,
+        protocol.min_reader_version(),
+    ) {
+        // NOTE: The table property is optional even when the feature is supported, and is allowed
+        // (but should be ignored) even when the feature is not supported. For details see
+        // https://github.com/delta-io/delta/blob/master/PROTOCOL.md#column-mapping
+        (Some(mode), 2) => mode,
+        (Some(mode), 3) if protocol.has_reader_feature(&ReaderFeatures::ColumnMapping) => mode,
         _ => ColumnMappingMode::None,
     }
 }
 
+/// When column mapping mode is enabled, verify that each field in the schema is annotated with a
+/// physical name and field_id; when not enabled, verify that no fields are annotated.
+pub fn validate_schema_column_mapping(schema: &Schema, mode: ColumnMappingMode) -> DeltaResult<()> {
+    if mode == ColumnMappingMode::Id {
+        // TODO: Support column mapping ID mode
+        return Err(Error::unsupported("Column mapping ID mode not supported"));
+    }
+
+    let mut validator = ValidateColumnMappings {
+        mode,
+        path: vec![],
+        err: None,
+    };
+    let _ = validator.transform_struct(schema);
```

Comment:
It feels weird to have to manually do error handling here and in all the `transform_*` methods. Why not make transform struct return a `Result<Option<...>>`?

Reply:
Tho ig this would require that we make *every* schema transform fallible 🤔

Reply:
I'd love a way to make this configurable, but I don't see any obvious way. Note that `Option` already poses similar problems, and it's quite common for transforms to either always return `None` (if they're just visiting the schema rather than changing it), or to always return `Some` (if they're modifying the schema but not filtering it in any way). In fact, there are even some that always return `Some(Cow::Owned)` because they unconditionally transform the schema.

If we wanted to capture this generically, we have three problems to solve:
1. What do the provided "leaf" methods return as their default"identify" transform? Today, it's `Some(Cow::Borrowed(val))`; `Cow::Borrowed(val).into()` would cover that and also cover `Cow<'a, T>` return type... but it does _not_ cover `DeltaResult<Cow<'a, T>>` (which has no blanket `impl From`).
2. How should the provided "internal" methods generically handle values that could be `None` or `Err`? `None` means "filter it out and keep going" while `Err` would mean "abort the traversal immediately"
3. How to handle the `?` operator, if the return type might be just `Cow<'a, T>`? Or `DeltaResult<Option<Cow<'a, T>>`?

To make matters worse, I can think of (at least) twelve return types the visitor might reasonably want to use:

<details>

1. `()` (visitor )
2.  `T` (unconditional transform)
3.  `Cow<'a, T>` (conditional transform)
4.  `DeltaResult<()>` (checker)
5.  `DeltaResult<T>` (fallible unconditional transform)
6.  `DeltaResult<Cow<'a, T>>` (fallible conditional transform)
7. `Option<&T>` (filter)
8. `Option<T>` (filtering unconditional transform)
9. `Option<Cow<'a, T>>` (filtering conditional transform)
10. `DeltaResult<Option<&T>>` (fallible filter)
11. `DeltaResult<Option<T>>` (fallible filtering unconditional transform)
12. `DeltaResult<Option<Cow<'a, T>>>` (fallible filtering conditional transform)

And that's ignoring the "aggregation" case where the visitor returns some completely other value instead of a schema (both fallible and infallible varieties). 

</details>

Note: If we _did_ decide to make the transforms fallible, the trait would need to know what error type to use. We could perhaps hard-wire it as `kernel::Error` but then infallible operations are out of luck. We can't default it because associated type defaults for traits are unstable rust. So we'd have to force every trait impl to define the type -- which is maybe not the end of the world -- and let infallible transforms specify `std::convert::Infallible` instead? But that has its own headaches when it's time to unpack the value (the `?` is still "conditional" and requires the calling function to return `Result`, and `unwrap` et al are still technically a panic risk. But at least the compiler recognizes that `let Ok(val) = returns_infallible_result()` is irrefutable in spite of not having an `else` clause.

Reply:
Woah this is a big design space. 
There seems to be two extremes going on here: 
1) Have a single design that we adapt to our usecases (current solution). More semantics is baked into the code. 
2) Have a type/trait that encompasses exactly the behaviour we want (all 12+ cases). In this case, the semantics are baked into the type system.

I wonder if we can strike a balance by having three variants:
- Cow<'a, T> to cover 2 and 3
- Option<Cow<'a, T>> to cover 1, 7, 8, 9
- DeltaResult<Option<Cow<'a, T>> to cover cases 10, 11, 12.
But then there will be semantics that aren't communicated by the type system. Maybe that's okay? For example:
- 2 will always return a `Cow::Owned`
- 7 will only return `None` or a `Some(Cow::Borrow)`
- 8 will only return `None` or `Some(Cow::Owned)` 


The idea of casting a wide net by introducing a generic error seems promising.
Regarding the generic defaulting: I could see a `SchemaTransform<E: Error>` with `type InfallibleSchemaTransform = SchemaTransform<Infallible>` and `type FallibleSchemaTransform = SchemaTransform<kernel::Error>`

This is definitely out of scope for this PR, but I think it's worth considering before implementing all the variety of transforms.

Reply:
Problem is, multiple slightly different types of transforms brings back the exact kind of code duplication the transforms were intended to eliminate...